### PR TITLE
[Pytorch] Add QuantizedTensor support in FusedAdam.step for MXFP8BlockScaling and Float8BlockScaling quantized model init.

### PR DESCRIPTION
**Pull request by jomitchellnv**
Summary                                                         

  - Add a QuantizedTensor code path in FusedAdam.step() that routes block-scaling quantized parameters (MXFP8, Float8Blockwise,
  NVFP4) through the FP32 Adam kernel on master weights, then writes back via quantize_() after the optimizer step
  - Add FSDP2 all-gather hooks (fsdp_pre_all_gather / fsdp_post_all_gather) to Float8BlockwiseQTensor, enabling end-to-end FSDP2
  training with Float8BlockScaling recipe
  - Add meta-device initialization pattern for FSDP2 + block-scaling quantized tensors (device="meta" + reset_parameters() after
  sharding), required because wrapper subclass tensors have data_ptr() == 0

  Details

  FusedAdam QuantizedTensor path (fused_adam.py):
  - After the existing Float8Tensor check (per-tensor FP8 → FP8 kernel), a new QuantizedTensor check catches block-scaling types
  and routes them to the FP32 master-weight Adam kernel
  - After all kernels run, a post-step loop calls local_p.quantize_(master_w.data) to write updated FP32 master weights back to
  both rowwise and columnwise quantized data
  - Raises RuntimeError if master_weights=False (required for block-scaling) or capturable=True (quantize writeback can't be
  CUDA-graph captured)

  Float8BlockwiseQTensor FSDP2 hooks (float8_blockwise_tensor.py):
  - fsdp_pre_all_gather: Transposes columnwise data/scales from (K, M) to (M, K) before all-gather so FSDP2's dim0 concatenation
  works correctly. Always sends both rowwise and columnwise (the GEMM kernel needs both forms since they have different shapes,
  unlike MXFP8 which stores both in the same shape)
  - fsdp_post_all_gather: Transposes columnwise back to (K, M), repads scale_inv to GEMM alignment (multiples of 4),
  constructs/updates the tensor
  - Adds aten.as_strided and aten.slice no-op handlers in __torch_dispatch__ to prevent FSDP2 from silently dequantizing the
  reconstructed tensor via QuantizedTensor's default dispatch
  - Only supports 2D block scaling (block_scaling_dim=2, the default for weights)

  Meta-device init for FSDP2 (run_fsdp2_fused_adam.py):
  - _build_model() accepts use_meta_device param; creates on device="meta" when fp8_init=True
  - _shard_model() detects meta params and calls reset_parameters() after fully_shard(), then restore_custom_attrs()
  - This is the only viable path for block-scaling QuantizedTensors with FSDP2 — direct CUDA init crashes in reset_sharded_param()
  due to inaccessible storage

  Tests:
  - Multi-GPU FSDP2: test_fsdp2_fused_adam_fp8_master_weights passes for DelayedScaling, Float8CurrentScaling, MXFP8BlockScaling,
  and Float8BlockScaling. NVFP4BlockScaling remains xfailed (no FSDP2 hooks)
  - Multi-GPU FSDP2: test_fsdp2_fused_adam_fp8_master_weights_no_meta documents the CUDA-init failure for block-scaling types
  (xfailed)
  - Single-GPU: TestFusedAdamMXFP8 and TestFusedAdamFloat8Block — each with a synthetic-grad master-weight test and a full
  forward/backward/step e2e test

  Test plan

  - pytest tests/pytorch/distributed/test_torch_fsdp2.py -v -k fused_adam (18 passed, 7 xfailed)
  - pytest tests/pytorch/test_fused_optimizer.py::TestFusedAdamMXFP8 -v (2 passed)
  - pytest tests/pytorch/test_fused_optimizer.py::TestFusedAdamFloat8Block -v (2 passed)
  - Verify no regressions in existing DelayedScaling / Float8CurrentScaling FSDP2 tests

 # Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 1]
```diff
@@ -622,6 +623,25 @@ def step(self, closure=None, grad_scaler=None):
...
+                            "FusedAdam without master_weights does not support "
+                            f"{type(local_p).__name__} parameters. Use master_weights=True."
+                        )
+                    # Route to the FP32 master-weight path: Adam updates the FP32 master,
+                    # then we write back to the quantized param after kernels run.
+                    p_f32_model.append(unscaled_state["master_param"].data)
+                    g_of_f32_model.append(p_grad.data)
+                    m_of_f32_model.append(unscaled_state["exp_avg"])
+                    v_of_f32_model.append(unscaled_state["exp_avg_sq"])
+                    quantized_params_to_update.append((p, unscaled_state["master_param"]))
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply transformer_engine/pytorch/optimizers/fused_adam.py:648]: added

File: tests/pytorch/distributed/run_fsdp2_fused_adam.py [hunk 2]
```diff
@@ -98,13 +110,22 @@ def _build_model(fp8_init, fuse_wgrad_accumulation=False, recipe=None):
...
     mesh = DeviceMesh("cuda", list(range(world_size)))
     for child in model.children():
         fully_shard(child, mesh=mesh)
     fully_shard(model, mesh=mesh)
     restore_custom_attrs(model, custom_attrs)
+    if has_meta_params:
+        for module in model.modules():
+            if hasattr(module, "reset_parameters"):
+                module.reset_parameters()
     return model
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply tests/pytorch/distributed/run_fsdp2_fused_adam.py:143]: I addressed this

**Comment by jomitchellnv** [comment]
@greptileai

File: tests/pytorch/distributed/run_fsdp2_fused_adam.py [hunk 3]
```diff
@@ -119,18 +148,11 @@ def test_fused_adam_fp8_master_weights(recipe=None):
     world_size, _, device = _setup()
 
     model = _build_model(fp8_init=True, recipe=recipe)
-
-    # Verify FP8 params created
-    qt_count = sum(1 for _, p in model.named_parameters() if isinstance(p, QuantizedTensor))
-    assert qt_count > 0, "No QuantizedTensor local tensors before training"
-
```
Comment (pstjohn) [inline tests/pytorch/distributed/run_fsdp2_fused_adam.py:126]: why delete?

Comment (jomitchellnv) [reply tests/pytorch/distributed/run_fsdp2_fused_adam.py:126]: the pre-sharding assertion was removed because with meta device init, params are meta tensors before sharding — they don't become QuantizedTensors until         
  reset_parameters() runs inside _shard_model(). The post-sharding assertion (line 162-167) still verifies QuantizedTensor local tensors exist after sharding completes.

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 4]
```diff
@@ -622,6 +623,25 @@ def step(self, closure=None, grad_scaler=None):
                     g_of_fp8_model.append(p_grad.data)
                     m_of_fp8_model.append(unscaled_state["exp_avg"])
                     v_of_fp8_model.append(unscaled_state["exp_avg_sq"])
+                elif isinstance(p, QuantizedTensor) or (
+                    isinstance(p, DTensor) and isinstance(p._local_tensor, QuantizedTensor)
+                ):
+                    # Block-scaling quantized params (MXFP8Tensor, Float8BlockwiseQTensor,
+                    # NVFP4Tensor). Operate on FP32 master weights, requantize back after
+                    # Adam update.
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/optimizers/fused_adam.py:631]: Minor Comment: can we add TODO
 Adam calculation and Quantization are unfused for now for these block-scaling parameters. Fusion needs to be done later.

Its just that the optimizer is called "fused adam" and we are indeed doing unfused adam for these 3 cases. So we know it needs to be done for all 3 of the cases.

File: transformer_engine/pytorch/tensor/float8_blockwise_tensor.py [hunk 5]
```diff
@@ -19,6 +19,11 @@
 from ._quantization_helpers import _IdentityFunc
 from ..utils import devices_match, round_up_to_nearest_multiple
 
+try:
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/tensor/float8_blockwise_tensor.py]: Is this dead code?

File: tests/pytorch/distributed/run_fsdp2_fused_adam.py (hunk 3 above)
Comment (vthumbe1503) [reply tests/pytorch/distributed/run_fsdp2_fused_adam.py:126]: @jomitchellnv I am curious, how was the test passing before then?

**Review by vthumbe1503 — APPROVED** [review]
Thanks for the PR. LGTM. Left a few minor questions/comments. Lets get this merged post CI sucess.

File: transformer_engine/pytorch/optimizers/fused_adam.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/optimizers/fused_adam.py:648]: @jomitchellnv this comment makes sense since grads can be bf16. Data can be misinterpreted in that case. Can we add this?

File: tests/pytorch/distributed/test_torch_fsdp2.py [hunk 6]
```diff
@@ -115,15 +115,34 @@ def _run_fused_adam_test(test_name, recipe="delayed_scaling"):
 
 @pytest.mark.skipif(NUM_PROCS < 2, reason="Requires 2+ GPUs")
 def test_fsdp2_fused_adam_fp8_master_weights(fp_recipe):
-    """FusedAdam(master_weights=True) + FSDP2 + quantized_model_init."""
-    if fp_recipe in ("Float8BlockScaling", "MXFP8BlockScaling", "NVFP4BlockScaling"):
+    """FusedAdam(master_weights=True) + FSDP2 + quantized_model_init (meta device init)."""
+    if fp_recipe in ("NVFP4BlockScaling",):
```
Comment (vthumbe1503) [inline tests/pytorch/distributed/test_torch_fsdp2.py:119]: Could you please add a pytest.skip  in case MXFP8/block-scaling support is missing? This test fails on H100 systems since it doesnt have MXFP8 suppport. The function to check support is already above I think. Its also done in a lot of other tests in the repo for you to refer.

File: tests/pytorch/test_fused_optimizer.py [hunk 7]
```diff
@@ -519,6 +520,259 @@ def test_fp8_model_weight_cast(self):
...
+        """quantized_model_init(Float8BlockScaling) + te.Linear + FusedAdam(master_weights=True).
+
+        Verifies:
+        - Model params are QuantizedTensors after init
+        - FP32 master weights track a reference Adam optimizer
+        - Params remain QuantizedTensors after training
+        """
+        model, recipe = self._build_model()
+
+        # Verify weight params are QuantizedTensors (bias stays bf16)
```
Comment (vthumbe1503) [inline tests/pytorch/test_fused_optimizer.py:684]: Could you also please add a pytest skip if float8 blockscaling is missing? I see you have added it already to mxfp8 test. We need it here similarly to avoid ci test failures.

File: transformer_engine/pytorch/optimizers/fused_adam.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply transformer_engine/pytorch/optimizers/fused_adam.py:648]: Good catch. The backward pass produces gradients in the compute dtype (typically BF16), but this path appends them to g_of_f32_model which feeds into the FP32 multi_tensor_adam kernel. Fixed by explicitly casting:                                       

`g_of_f32_model.append(p_grad.data.float())`

This matches the intent — master weights, exp_avg, and exp_avg_sq are all FP32, so the gradients should be too. The existing per-tensor Float8Tensor path doesn't need this because it uses a dedicated mixed-precision FP8 kernel (multi_tensor_adam_fp8), but our QuantizedTensor path routes through the plain FP32 kernel.

File: transformer_engine/pytorch/optimizers/fused_adam.py (hunk 4 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/optimizers/fused_adam.py:631]: Added

File: tests/pytorch/distributed/run_fsdp2_fused_adam.py (hunk 3 above)
Comment (jomitchellnv) [reply tests/pytorch/distributed/run_fsdp2_fused_adam.py:126]: Before (on main): `_build_model(fp8_init=True)` created the model on CUDA. With quantized_model_init, parameters were immediately QuantizedTensors. So the pre-sharding check passed:                                                                         
```python
qt_count = sum(1 for _, p in model.named_parameters() if isinstance(p, QuantizedTensor))
assert qt_count > 0, "No QuantizedTensor local tensors before training"                                                                                                                                                                                     
```
                                                                  
After (our changes): We switched to device="meta" to fix the FSDP2 crash on wrapper subclass tensors with data_ptr() == 0. With meta-device init, parameters are meta tensors after _build_model() — they only become QuantizedTensors after _shard_model() calls reset_parameters(). So the pre-sharding assertion was no longer valid and was removed.

The assertion passed originally — we removed it because we changed the initialization strategy, making it structurally impossible to satisfy. The equivalent check now lives post-sharding (lines 159-167).

File: transformer_engine/pytorch/tensor/float8_blockwise_tensor.py (hunk 5 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/tensor/float8_blockwise_tensor.py]: yes ill remove it.

File: tests/pytorch/test_fused_optimizer.py (hunk 7 above)
Comment (jomitchellnv) [reply tests/pytorch/test_fused_optimizer.py:684]: Done — added te.is_fp8_block_scaling_available(return_reason=True) check in setup_method and pytest.skip guards in both test_float8block_linear_fused_adam_master_weights and test_float8block_linear_forward_backward_step, matching the MXFP8 pattern.

File: tests/pytorch/distributed/test_torch_fsdp2.py (hunk 6 above)
Comment (jomitchellnv) [reply tests/pytorch/distributed/test_torch_fsdp2.py:119]: done!. I changed `_parametrize_fp8_recipes()` to use `pytest.mark.skipif` instead of `pytest.mark.xfail` for hardware support checks. On systems without MXFP8/block-scaling support, tests are now skipped outright instead of running the torchrun subprocess and failing noisily. Per-test xfail marks for known functional gaps (e.g., NVFP4 lacking FSDP2 hooks) remain unchanged since those are feature limitations, not hardware issues.

File: tests/pytorch/distributed/run_fsdp2_fused_adam.py (hunk 3 above)
Comment (vthumbe1503) [reply tests/pytorch/distributed/run_fsdp2_fused_adam.py:126]: Got it. Thanks!

File: tests/pytorch/distributed/test_torch_fsdp2.py [hunk 8]
```diff
@@ -189,13 +214,18 @@ def test_fsdp2_dcp_output_parity_async(fp_recipe):
             "multi_tensor_apply: CUDA Error: an illegal memory access was encountered"
         )
 
-    if fp_recipe == "Float8BlockScaling" and torch.cuda.get_device_capability()[0] == 12:
+    if fp_recipe == "NVFP4BlockScaling":
```
Comment (jomitchellnv) [inline tests/pytorch/distributed/test_torch_fsdp2.py:217]: the previous code was xfailing this only on sm120... it should be more general
